web/src/copy_and_paste.js

```diff
@@ -293,29 +293,138 @@ function get_end_tr_from_endc($endc) {
+        return node.getAttribute("alt");
+    }
+    const src = node.getAttribute("src") || node.getAttribute("href") || "";
+    const title = cleanAttribute(node.getAttribute("title")) || "";
+    // Using Zulip's link like syntax for images
+    return src ? "[" + title + "](" + src + ")" : node.getAttribute("alt") || "";
+}
+
 export function paste_handler_converter(paste_html) {
-    const turndownService = new TurndownService();
+    // turning off escaping (for now) to remove extra `/`
+    TurndownService.prototype.escape = (string) => string;
+
+    const turndownService = new TurndownService({emDelimiter: "*", codeBlockStyle: "fenced"});
     turndownService.addRule("headings", {
         filter: ["h1", "h2", "h3", "h4", "h5", "h6"],
         replacement(content) {
             return content;
         },
     });
-    turndownService.addRule("emphasis", {
-        filter: ["em", "i"],
+    turndownService.addRule("strikethrough", {
+        filter: ["del", "s", "strike"],
         replacement(content) {
-            return "*" + content + "*";
+            return "~~" + content + "~~";
         },
     });
-    // Checks for raw links without custom text or title.
     turndownService.addRule("links", {
+        filter: ["a"],
+        replacement(content, node) {
+            if (node.href === node.innerHTML) {
+                // Checks for raw links without custom text.
+                return content;
+            }
+            if (node.childNodes.length === 1 && node.firstChild.nodeName === "IMG") {
+                // ignore link's url if it only has an image
+                return content;
+            }
+            return "[" + content + "](" + node.href + ")";
+        },
+    });
+    turndownService.addRule("listItem", {
+        filter: "li",
+        replacement(content, node) {
+            content = content
+                .replace(/^\n+/, "") // remove leading newlines
+                .replace(/\n+$/, "\n") // replace trailing newlines with just a single one
+                .replace(/\n/gm, "\n  "); // custom 2 space indent
+            let prefix = "* ";
+            const parent = node.parentNode;
+            if (parent.nodeName === "OL") {
+                const start = parent.getAttribute("start");
+                const index = Array.prototype.indexOf.call(parent.children, node);
+                prefix = (start ? Number(start) + index : index + 1) + ". ";
+            }
+            return prefix + content + (node.nextSibling && !/\n$/.test(content) ? "\n" : "");
+        },
+    });
+    turndownService.addRule("zulipCodeBlock", {
+        filter(node, options) {
+            return (
+                options.codeBlockStyle === "fenced" &&
+                node.nodeName === "CODE" &&
+                node.parentElement?.nodeName === "PRE"
+            );
+        },
+
+        replacement(content, node, options) {
+            const language = node.closest(".codehilite")?.dataset?.codeLanguage || "";
```

Comment:
This filter seems to apply to non-Zulip code blocks as well; is that expected? 

I guess the `data-code-language` feature is a Zulip-specific attribute, and maybe there's nothing we can do with third-party code blocks to get a `pygments` language out of a code block, but we should probably document what we're doing here in comments.

Reply:
This is meant to apply to Zulip code blocks specifically, but I wasn't able to find a way to uniquely and completely identify them -- `data-code-language` is only present for code blocks with an explicitly specified language, so including that in the filter will leave out zulip code blocks with no specified language.

But also, I can't think of any downsides of this applying to non Zulip code blocks unintentionally 🤔

Reply:
OK